src/migration/cassandra/impl/FullTableScanner.hpp

```diff
@@ -0,0 +1,175 @@
+#include <cstddef>
+#include <cstdint>
+#include <limits>
+#include <ranges>
+#include <vector>
+
+namespace migration::cassandra::impl {
+
+/**
+ * @brief The token range used to split the full table scan into multiple ranges.
+ */
+struct TokenRange {
+    std::int64_t start;
+    std::int64_t end;
+
+    /**
+     * @brief Construct a new Token Range object
+     *
+     * @param start The start token
+     * @param end The end token
+     */
+    TokenRange(std::int64_t start, std::int64_t end) : start{start}, end{end}
+    {
+    }
+};
+
+/**
+ * @brief The concept for an adapter.
+ */
+template <typename T>
+concept CanReadByTokenRange = requires(T obj, TokenRange const& range, boost::asio::yield_context yield) {
+    { obj.readByTokenRange(range, yield) } -> std::same_as<void>;
+};
+
+/**
+ * @brief The full table scanner. It will split the full table scan into multiple ranges and read the data in given
+ * executor.
+ *
+ * @tparam TableAdapter The table adapter type
+ */
+template <CanReadByTokenRange TableAdapter>
+class FullTableScanner {
+    /**
+     * @brief The helper to generate the token ranges.
+     */
+    struct TokenRangesProvider {
+        uint32_t numRanges_;
+
+        TokenRangesProvider(uint32_t numRanges) : numRanges_{numRanges}
+        {
+        }
+
+        [[nodiscard]] std::vector<TokenRange>
+        getRanges() const
+        {
+            auto const minValue = std::numeric_limits<std::int64_t>::min();
+            auto const maxValue = std::numeric_limits<std::int64_t>::max();
+            if (numRanges_ == 1)
+                return {TokenRange{minValue, maxValue}};
+
+            // Safely calculate the range size using uint64_t to avoid overflow
+            uint64_t rangeSize = (static_cast<uint64_t>(maxValue) * 2) / numRanges_;
+
+            std::vector<TokenRange> ranges;
+            ranges.reserve(numRanges_);
+
+            for (std::int64_t i = 0; i < numRanges_; ++i) {
+                int64_t start = minValue + i * rangeSize;
+                int64_t end = (i == numRanges_ - 1) ? maxValue : start + static_cast<int64_t>(rangeSize) - 1;
+                ranges.emplace_back(start, end);
+            }
+
+            return ranges;
+        }
+    };
+
+    [[nodiscard]] auto
+    spawnWorker()
+    {
+        return ctx_.execute([this](auto token) {
+            while (not token.isStopRequested()) {
+                auto cursor = queue_.tryPop();
+                if (not cursor.has_value()) {
+                    return;  // queue is empty
+                }
+                reader_.readByTokenRange(cursor.value(), token);
+            }
+        });
+    }
+
+    void
+    load(size_t workerNum)
+    {
+        namespace vs = std::views;
+
+        tasks_.reserve(workerNum);
+
+        for ([[maybe_unused]] auto taskId : vs::iota(0u, workerNum))
+            tasks_.push_back(spawnWorker());
+    }
+
+    static auto constexpr CursorNumMultiplier = 100;  // will generate 100*workersNum cursors
```

Comment:
Is this an arbitrary number? how was it chosen? Should not be configurable in any way?

Reply:
Because the total cursor numbers is `workersNum* CursorNumMultiplier`, workersNum is configurable coroutine numbers. Each coroutine can handle `CursorNumMultiplier` cursor by average. But the content are not distribute evenly, so `CursorNumMultiplier` split the cursors range to smaller piece. It is hard to decide which number is the best without the insight of data distribution in particular tables. But I think with CursorNumMultiplier 100, user can tweet total cursors number by adjusting the number of coroutines.